### ISSUE TITLE
Not an issue, even less advertisement, more code update...

### ISSUE THREAD

**Issue by zelogik** [MikeBland/SbusToPpm#9]
Hi @MikeBland,

First, thanks for your optimized code (I don't like digitalWrite, pinMode, delay/millis, in arduino code too :-D).

I have make a fork on https://github.com/zelogik/SbusToPWM.

Only code cleaning for the moment, some code factorization on the go, but mainly AVR architecture compatibility in the future (new ATTiny1616 for example).

So the question, as the code seem depreciated, do you have time for merge request, or do you prefer a completely separate fork/repo ? 

thanks :D

**Comment by MikeBland** [comment]
The code uses tabs, set to 2. The editor I use does this automatically and the code is "clean" if you have tabs set to 2.

Stay on a separate fork I think.

Not a bug in delay(), lms is counted down to 0, 0 is not >0.

**Comment by zelogik** [comment]
Yes for the bug that is not a bug, you can even see a '?' because I was not sure about it 5hours ago, and delay() is not used in the code so have been removed.

And I will stay on separate fork, to much modification on my last pull, and something 'easily' readable for me is not necessary the same for everybody.
ie: I prefer for loop, than many almost repetitive lines :D, normally speed should be the same (need to be tested), like in the setPulseTimes() code.

For the tab vs space, no problem, just VSCode that show some problems with your raw code: mix between tab and space.
![image](https://github.com/user-attachments/assets/acc40d24-9792-4aa0-ab60-bab671f35e75)  ![image](https://github.com/user-attachments/assets/21b92f0e-bb92-486b-84d6-c4e9ca435bc6)  ![image](https://github.com/user-attachments/assets/622d015a-9c48-4aa6-abdb-4dabb4148a03)

And for the space before each ;, I try to use the K&R style, too much space make my eyes hurt :p (no offense :p)

Many critics, but many parts of your code is nice (mainly the input pulse logic) and really useful.
Thanks

**Comment by zelogik** [comment]
and I try to optimize the flash size code, so I could put it on something smaller than a 328:

Before lifting:
Sketch uses 3218 bytes (10%) of program storage space. Maximum is 30720 bytes.
Global variables use 229 bytes (11%) 

After lifting:
Sketch uses 2594 bytes (8%) of program storage space. Maximum is 30720 bytes.
Global variables use 228 bytes (11%) of dynamic memory, leaving 1820 bytes for local variables. Maximum is 2048 bytes.

Hard to remove global variables, code is well programmed here :D

**Comment by zelogik** [comment]
@MikeBland 
As you not far, what the use of sortToCopy() ? there is no reference.

**Comment by MikeBland** [comment]
Take care with your changes.
Because the sketch has a main() function, this overrides any normal arduino code. For example, setup() and init() need to be called from main(), they are not called from anywhere else. There is no normal arduino code involved, everything is in the single .ino file.
Using a loop in setPulseTimes() is likely to be slower. Pulses[5].port is accessed using direct addressing, Pulses[4+i].port will need the address calculated, then referenced using indirect addressing.
Declaring variables as static instead of global doesn't gain any code or data space and only hides those variables from the map file, so you don't know where they are stored. Knowing  where everything is stored is useful when debugging. If, for example, a bug causes a memory access beyond the end of an array, it is useful to know what variables might have been changed.

You appear to be making LastPulsesStartTime static in two, different functions, yet it should be global so it is common to both functions.
in main() you are declaring y as a uint32 instead of uint16. uint16 is the required size to be used with rate and LastPulsesStartTime (both uint16). Declaring y as uint32 will cause everything to be promoted to uint32 resulting in significantly more code to do arithmetic and comparisons as well as making an incorrect decision.

I'm not sure your calling setBytes() from main, in a loop, is following the correct sequence. Having started the first 4 pulses, my main continues to loop and then starts the next 4 pulses 3mS after the previous 4 pulses.

sortToCopy() is not needed. The compiler/linker actually removes both that and delay().

**Comment by zelogik** [comment]
Thanks for your reply/comment.
- yes, replacing main() by loop() is useful there for me, 'lose' C/C++ code 'compatibility' for a more arduino code specific...:
  - pros:
     - setup() is called one time without any call in the loop()
  - cons:
     - millis()/micros()/delay() are added to the program and add approx 300bytes, but as I want to use attiny serie 1 and use megatinyCore, it should be ok as I can remove it when compiling.
- So question is: should I continu to use main() and set an init variable as the main will be completely modified, or get loop() with the cons of increasing flash size...

- for the loop in setPulseTimes(), yes I'm aware that some cpu cycles are 'lost' with for/if/i+x, need to be tested to see if it's really slower and problematic or not for that function, but from my point of view, the 'randomly' call to checkInput() annoy me more.
- put static vs global, it's just for limiting scope to the function, and don't use it on two different functions with unintended variable access, and create bug/strange behavior. By experience, global variable are harder to debug/view change, when used...
- LastPulsesStartTime() in both function, yes you right, will be corrected, loop/main function need big change.
- For y as uint32_t, it's was intentional, as i take from uint32_t micros()... So seem you can lose bit in your code uint32_t -> uint16_t ...
- the ~~main~~ loop() function need to be modified, and the call to setBytes is here, just to be improved/debugged, that nested if (7 nested if ...), annoy me too. Just need to understand better your code.
- I prefer to remove/comment not used code for readability, even if compiler/linker remove it.

So yes, keep separate repo for your and mine seem a good choice, as we don't code the same way. :-D
